src/frontend/src/scheduler/hummock_snapshot_manager.rs

```diff
@@ -13,125 +13,215 @@
 // limitations under the License.
 
 use std::collections::{BTreeMap, HashSet};
+use std::sync::atomic::{AtomicU64, Ordering};
 use std::sync::Arc;
+use std::time::{Duration, Instant};
 
+use anyhow::anyhow;
 use log::error;
-use tokio::sync::Mutex;
+use tokio::sync::mpsc::{channel, Sender};
+use tokio::sync::oneshot::{channel as once_channel, Sender as Callback};
 
 use crate::meta_client::FrontendMetaClient;
 use crate::scheduler::plan_fragmenter::QueryId;
-use crate::scheduler::SchedulerError::PinSnapshot;
-use crate::scheduler::SchedulerResult;
+use crate::scheduler::{SchedulerError, SchedulerResult};
+
+const MAX_WAIT_EPOCH_REQUEST_NUM: usize = 4096;
+const UNPIN_INTERVAL_SECS: u64 = 10;
```

Comment:
10 secs looks too long for unpin, given that our epoch update at 100 ms freq.

An strategy is to unpin when we recive a epoch update.

Reply:
10 is short enough. Because almost all our compaction jobs can run a long time more than 10 seconds. It means that all data will cost at several tens second from shared-buffer to level-1 in S3.